homeassistant/components/proxmoxve/__init__.py

```diff
@@ -71,52 +83,193 @@
+                    host, port, user, realm, password, verify_ssl
+                )
+                proxmox_client.build_client()
+            except AuthenticationError:
+                _LOGGER.warning(
+                    "Invalid credentials for proxmox instance %s:%d", host, port
+                )
+                continue
+            except SSLError:
+                _LOGGER.error(
+                    'Unable to verify proxmox server SSL. Try using "verify_ssl: false"'
+                )
+                continue
+
+            return proxmox_client
+
+    proxmox_client = await hass.async_add_executor_job(build_client)
+
+    async def async_update_data() -> dict:
+        """Fetch data from API endpoint."""
+
+        proxmox = proxmox_client.get_api_client()
+
+        def poll_api() -> dict:
+            data = {}
+
+            for host_config in config[DOMAIN]:
+                host_name = host_config["host"]
+
+                data[host_name] = {}
+
+                for node_config in host_config["nodes"]:
+                    node_name = node_config["node"]
+                    data[host_name][node_name] = {}
+
+                    for vm_id in node_config["vms"]:
+                        data[host_name][node_name][vm_id] = {}
+
+                        vm_status = call_api_container_vm(
+                            proxmox, node_name, vm_id, TYPE_VM
+                        )
+
+                        if vm_status is None:
+                            _LOGGER.warning("Vm/Container %s unable to be found", vm_id)
+                            data[host_name][node_name][vm_id] = None
+                            continue
+
+                        data[host_name][node_name][vm_id] = parse_api_container_vm(
+                            vm_status
+                        )
 
-        hass.data[PROXMOX_CLIENTS][f"{host}:{port}"] = proxmox_client
+                    for container_id in node_config["containers"]:
+                        data[host_name][node_name][container_id] = {}
 
-    if hass.data[PROXMOX_CLIENTS]:
-        hass.helpers.discovery.load_platform(
-            "binary_sensor", DOMAIN, {"entries": config[DOMAIN]}, config
+                        container_status = call_api_container_vm(
+                            proxmox, node_name, container_id, TYPE_CONTAINER
+                        )
+
+                        if container_status is None:
+                            _LOGGER.error(
+                                "Vm/Container %s unable to be found", container_id
+                            )
+                            data[host_name][node_name][container_id] = None
+                            continue
+
+                        data[host_name][node_name][
+                            container_id
+                        ] = parse_api_container_vm(container_status)
+
+            return data
+
+        return await hass.async_add_executor_job(poll_api)
+
+    coordinator = DataUpdateCoordinator(
```

Comment:
I think that it would make more sense to create an update coordinator per instance. That way the updates will run in parallel. It also means that if a user disables all entities of a single instance, no data will be fetched.